pruning/manager.go

```diff
@@ -0,0 +1,282 @@
+package pruning
+
+import (
+	"container/list"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/cosmos/cosmos-sdk/pruning/types"
+)
+
+// Manager is an abstraction to handle the logic needed for
+// determinging when to prune old heights of the store
+// based on the strategy described by the pruning options.
+type Manager struct {
+	db               dbm.DB
+	logger           log.Logger
+	opts             types.PruningOptions
+	snapshotInterval uint64
+	// Although pruneHeights happen in the same goroutine with the normal execution,
+	// we sync access to them to avoid soundness issues in the future if concurrency pattern changes.
+	pruneHeightsMx         sync.Mutex
+	pruneHeights           []int64
+	// Snapshots are taken in a separate goroutine from the regular execution
+	// and can be delivered asynchrounously via HandleHeightSnapshot.
+	// Therefore, we sync access to pruneSnapshotHeights with this mutex. 
+	pruneSnapshotHeightsMx sync.Mutex
+	// These are the heights that are multiples of snapshotInterval and kept for state sync snapshots.
+	// The heights are added to this list to be pruned when a snapshot is complete.
+	pruneSnapshotHeights   *list.List
+}
+
+// NegativeHeightsError is returned when a negative height is provided to the manager.
+type NegativeHeightsError struct {
+	Height int64
+}
+
+var _ error = &NegativeHeightsError{}
+
+func (e *NegativeHeightsError) Error() string {
+	return fmt.Sprintf("failed to get pruned heights: %d", e.Height)
+}
+
+var (
+	pruneHeightsKey         = []byte("s/pruneheights")
+	pruneSnapshotHeightsKey = []byte("s/pruneSnheights")
```

Comment:
```suggestion
	pruneSnapshotHeightsKey = []byte("s/prunesnapshotheights")
```

Reply:
Btw, these are new keys, correct?

Reply:
`s/pruneheights` is old and `"s/pruneSnheights"` is new. However, `s/pruneSnheights` is already present in Osmosis.